### PR TITLE
feat: implement SerializableModuleElaborator

### PR DESCRIPTION
**Pull request by unlsycn**
This PR implements the SerializableModuleElaborator trait, which is used to create elaborators for SerializableModule. In t1, we compile all the Scala code into a jar file and specify different main classes to elaborate different modules. 
The config API is used to guarantee the type safety of the parameters, and the design API actually produces the circuit, which is used this way:
```scala
  @main
  def config(
    @arg(name = "parameter") parameter:  GCDParameterMain,
    @arg(name = "target-dir") targetDir: os.Path = os.pwd
  ) =
    os.write.over(targetDir / s"${topName}.json", configImpl(parameter.convert))

  @main
  def design(
    @arg(name = "parameter") parameter:  os.Path,
    @arg(name = "target-dir") targetDir: os.Path = os.pwd
  ) = {
    val (firrtl, annos) = designImpl[GCD, GCDParameter](os.read.stream(parameter))
    os.write.over(targetDir / s"${topName}.fir", firrtl)
    os.write.over(targetDir / s"${topName}.anno.json", annos)
  }
```

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash

#### Release Notes

Add SerializableModuleElaborator

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
+package chisel3.experimental.util
```
Comment (sequencer) [inline src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala:1]: Please add Apache license header.

File: src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import scala.reflect.runtime.universe.{runtimeMirror, typeOf}
+
+/** The target design emitted by [[SerializableModuleElaborator]], which contains
+  * @param fir the firrtl circuit
+  * @param annos annotations which is serializable
+  */
+case class SerializableModuleDesign(fir: firrtl.ir.Circuit, annos: Seq[firrtl.annotations.Annotation]) {
+  def firFile = fir.serialize
+  def annosFile = firrtl.annotations.JsonProtocol.serializeRecover(annos)
+}
```
Comment (sequencer) [inline src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala:16]: How about directly introducing the tuple for result? Adding new API here for only one usage might be over design, and I don't like to getting it deprecate in the future.

**Comment by unlsycn** [comment]
> How about directly introducing the tuple for result? Adding new API here for only one usage might be over design, and I don't like to getting it deprecate in the future.

I was using it originally to provide unserialized structures to the user. Now we return a tuple of Writable, which should be used in this way:
```scala
val (firrtl, annos) = designImpl[MySerializableModule, MySerializableModuleParameter](os.read(parameter))
os.write.over(os.pwd / "GCD.fir", firrtl)
os.write.over(os.pwd / "GCD.anno.json", annos)
```

File: src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala (hunk 2 above)
Comment (unlsycn) [reply src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala]: I used it originally to provide unserialized structures to user. Now we return a tuple of geny.Readable hence we can use it in this way:
```scala
val (firrtl, annos) = designImpl[MySerializableModule, MySerializableModuleParameter](os.read.stream(parameter))
os.write.over(os.pwd / "GCD.fir", firrtl)
os.write.over(os.pwd / "GCD.anno.json", annos)
```

File: src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      ): firrtl.AnnotationSeq
+    ) { case (annos, stage) => stage.transform(annos) }
+      .flatMap {
+        case firrtl.stage.FirrtlCircuitAnnotation(circuit) =>
+          fir = circuit
+          None
+        case _: firrtl.options.Unserializable => None
+        case a => Some(a)
+      }
+    val firrtlStream: Readable = fir.serialize
```
Comment (sequencer) [inline src/main/scala/chisel3/experimental/util/SerializableModuleElaborator.scala:73]: I still doubt there is an extra copy to Sting here, but the return type is good now and we can improve the internal performance with StringBuild in the future.

**Review by sequencer — APPROVED** [review]
Maybe you or someone can add more documents in the documentation on how to use the SerializableModule in the future, I think for now the chisel-nix project already give a comprehensive example to it. Please update that to use this Elaborator, and point new users to there.

**Pull request merged**